### PR TITLE
Use new Kafka API for consuming messages

### PR DESCRIPTION
**Pull request by kokosing**
Use new Kafka API for consuming messages

Previously we created one Presto split per one segment file in each
Kafka partitions.
This approach creates one split per each Kafka partition. This may
possibly limit the concurrency where there is low number of Kafka
brokers and splits could be processed much longer.
However, this follows best practices of Kafka messages distribution
and also allows Presto to leverage other Kafka APIs features that
were not possible to use with SimpleConsumer.

### CODE REVIEW

**Comment by kokosing** [comment]
Depends on https://github.com/prestosql/presto/pull/2302

**Comment by kokosing** [comment]
@findepi ping

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java [hunk 1]
```diff
@@ -16,6 +16,7 @@
 import com.google.common.collect.ImmutableList;
```
Comment (sopel39) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/KafkaQueryRunner.java]: Add more rationale to commit message

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        nodes = kafkaConfig.getNodes();
+        kafkaBufferSize = kafkaConfig.getKafkaBufferSize();
+        this.classLoader = requireNonNull(classLoader, "classLoader is null");
+    }
+
+    public KafkaConsumer<byte[], byte[]> create()
+    {
+        Properties properties = new Properties();
+        properties.put(
+                ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java]: static import

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                nodes.stream()
+                        .map(HostAddress::toString)
+                        .collect(joining(",")));
+        properties.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        properties.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        properties.put(ConsumerConfig.RECEIVE_BUFFER_CONFIG, Long.toString(kafkaBufferSize.toBytes()));
+        properties.put(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, Boolean.toString(false));
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(classLoader)) {
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java]: is that needed? Use `ClassLoaderSafeConnector*` instead?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java [hunk 4]
```diff
@@ -178,31 +128,4 @@ private static boolean isURI(String location)
         }
         return true;
     }
-
-    private static long[] findAllOffsets(SimpleConsumer consumer, String topicName, int partitionId)
```
Comment (kokosing) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java]: Try to use `org.apache.kafka.clients.consumer.KafkaConsumer#beginningOffsets` to restore segments

**Comment by kokosing** [comment]
@sopel39 I added a configuration property `kafka.messages-per-split` which defaults to `100_000`. So in case of big partitions they can be still processed in parallel. By default segment size is set to `1GB` so one split per segment didn't sound like a good idea. Now user has a control about split granularity.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConnector.java [hunk 5]
```diff
@@ -31,16 +29,15 @@
         implements Connector
 {
     private final LifeCycleManager lifeCycleManager;
-    private final KafkaMetadata metadata;
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConnector.java]: undo?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConnectorFactory.java [hunk 6]
```diff
@@ -81,6 +86,10 @@ public Connector create(String catalogName, Map<String, String> config, Connecto
                 .setRequiredConfigurationProperties(config)
                 .initialize();
 
-        return injector.getInstance(KafkaConnector.class);
+        return new KafkaConnector(
+                injector.getInstance(LifeCycleManager.class),
+                injector.getInstance(KafkaMetadata.class),
+                new ClassLoaderSafeConnectorSplitManager(injector.getInstance(KafkaSplitManager.class), classLoader),
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConnectorFactory.java]: you can just use `KafkaConnectorFactory` classloader

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 7]
```diff
@@ -138,57 +130,30 @@ public Type getType(int field)
...
-
-                    if (messageOffset >= cursorOffset) {
-                        return nextRow(currentMessageAndOffset);
-                    }
+            if (!records.hasNext()) {
+                if (kafkaConsumer.position(topicPartition) >= split.getMessages().getEnd()) {
+                    return false;
                 }
-                messageAndOffsetIterator = null;
+                records = kafkaConsumer.poll(100).iterator();
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: extract 100 as static final. How was 100 chosen?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 8]
```diff
@@ -138,57 +130,30 @@ public Type getType(int field)
...
-            return false;
-        }
-
-        private boolean nextRow(MessageAndOffset messageAndOffset)
-        {
-            cursorOffset = messageAndOffset.offset() + 1; // Cursor now points to the next message.
-            totalBytes += messageAndOffset.message().payloadSize();
-            totalMessages++;
+            requireNonNull(message, "message is null");
+            completedBytes += max(message.serializedKeySize(), 0) + max(message.serializedValueSize(), 0);
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: can `serializedKeySize` be negative?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.joining;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.RECEIVE_BUFFER_CONFIG;
+import static org.apache.kafka.clients.consumer.ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG;
+
+public class KafkaConsumerFactory
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaConsumerFactory.java]: rename to `KafkaConsumerManager` or change variables for `consumerFactory`

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java [hunk 10]
```diff
@@ -303,47 +259,12 @@ private void checkFieldType(int field, Class<?> expected)
...
-                                "Cannot read data from topic '%s', partition '%s', startOffset %s, endOffset %s, leader %s ",
-                                split.getTopicName(),
-                                split.getPartitionId(),
-                                split.getStart(),
-                                split.getEnd(),
-                                split.getLeader()),
-                        e);
-            }
-        }
+    private static <T> Iterator<T> emptyIterator()
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaRecordSet.java]: use https://guava.dev/releases/15.0/api/docs/com/google/common/collect/Iterators.html#emptyIterator()

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplit.java [hunk 11]
```diff
@@ -41,8 +34,7 @@
     private final Optional<String> keyDataSchemaContents;
     private final Optional<String> messageDataSchemaContents;
     private final int partitionId;
-    private final long start;
-    private final long end;
+    private final Range messages;
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplit.java]: rename to `messagesRange`

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java [hunk 12]
```diff
@@ -61,66 +51,51 @@
...
+            Map<TopicPartition, Long> partitionEndOffsets = kafkaConsumer.endOffsets(topicPartitions);
+
+            ImmutableList.Builder<KafkaSplit> splits = ImmutableList.builder();
+            Optional<String> keyDataSchemaContents = kafkaTableHandle.getKeyDataSchemaLocation()
+                    .map(KafkaSplitManager::readSchema);
+            Optional<String> messageDataSchemaContents = kafkaTableHandle.getMessageDataSchemaLocation()
+                    .map(KafkaSplitManager::readSchema);
+            for (PartitionInfo partitionInfo : partitionInfos) {
+                TopicPartition topicPartition = toTopicPartition(partitionInfo);
+                HostAddress leader = HostAddress.fromParts(partitionInfo.leader().host(), partitionInfo.leader().port());
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java]: are there more nodes containing topic? For large topics we could distribute load such way

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java [hunk 13]
```diff
@@ -131,6 +106,11 @@ public ConnectorSplitSource getSplits(ConnectorTransactionHandle transaction, Co
         }
     }
 
+    private static TopicPartition toTopicPartition(PartitionInfo partitionInfo)
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/KafkaSplitManager.java]: How does `TopicPartition` and `PartitionInfo` differ?

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    @JsonProperty
+    public long getBegin()
+    {
+        return begin;
+    }
+
+    @JsonProperty
+    public long getEnd()
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java]: write `boolean contains(long position)` instead.

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    {
+        return end;
+    }
+
+    public Stream<Range> partition(int partitionSize)
+    {
+        return Streams.stream(() -> new AbstractIterator<Range>()
+        {
+            private long position = Range.this.begin;
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java]: `Range.this.` is redundant

File: presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+        return begin;
+    }
+
+    @JsonProperty
+    public long getEnd()
+    {
+        return end;
+    }
+
+    public Stream<Range> partition(int partitionSize)
```
Comment (sopel39) [inline presto-kafka/src/main/java/io/prestosql/plugin/kafka/Range.java]: why not return plain here and use basic loop?

File: presto-kafka/src/test/java/io/prestosql/plugin/kafka/util/JsonSerializer.java [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+ */
+package io.prestosql.plugin.kafka.util;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.apache.kafka.common.serialization.Serializer;
+
+import java.util.Map;
+
+public class JsonSerializer
```
Comment (sopel39) [inline presto-kafka/src/test/java/io/prestosql/plugin/kafka/util/JsonSerializer.java]: why this is needed?

File: presto-spi/src/main/java/io/prestosql/spi/connector/classloader/ClassLoaderSafeConnectorRecordSetProvider.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+import io.prestosql.spi.connector.ConnectorSplit;
+import io.prestosql.spi.connector.ConnectorTableHandle;
+import io.prestosql.spi.connector.ConnectorTransactionHandle;
+import io.prestosql.spi.connector.RecordSet;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class ClassLoaderSafeConnectorRecordSetProvider
```
Comment (sopel39) [inline presto-spi/src/main/java/io/prestosql/spi/connector/classloader/ClassLoaderSafeConnectorRecordSetProvider.java]: I would extract this to separate commit

File: presto-spi/src/test/java/io/prestosql/spi/connector/classloader/TestClassLoaderSafeConnectorRecordSetProvider.java [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package io.prestosql.spi.connector.classloader;
+
+import io.prestosql.spi.connector.ConnectorRecordSetProvider;
+import org.testng.annotations.Test;
+
+import static io.prestosql.spi.testing.InterfaceTestUtils.assertAllMethodsOverridden;
+
+public class TestClassLoaderSafeConnectorRecordSetProvider
```
Comment (sopel39) [inline presto-spi/src/test/java/io/prestosql/spi/connector/classloader/TestClassLoaderSafeConnectorRecordSetProvider.java]: you don' test `ClassLoaderSafeRecordSet`
